core/unix/memquery_test.h

```diff
@@ -0,0 +1,311 @@
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/*
+ * memquery_test.h - unit test for unix memquery logic, it is meant to be
+ * included into memquery.c after its includes.
+ *
+ * See also memquery_test_data.h
+ */
+
+#define MEMQUERY_ITERATOR_START memquery_iterator_start_test
+#define MEMQUERY_ITERATOR_NEXT memquery_iterator_next_test
+#define MEMQUERY_ITERATOR_STOP memquery_iterator_stop_test
+#define MODULE_IS_HEADER module_is_header_test
+#define MODULE_WALK_PROGRAM_HEADERS module_walk_program_headers_test
+
+typedef struct {
+    memquery_iter_t iter_result;
+    bool is_header;
+    app_pc mod_base, mod_end;
+} fake_memquery_result;
+
+typedef struct {
+    fake_memquery_result *iters;
+    size_t iters_count;
+
+    const char *in_name;
+    app_pc in_start;
+
+    int want_return;
+    app_pc want_start;
+    app_pc want_end;
+    const char *want_fulldir;
+    const char *want_filename;
+} memquery_library_bounds_test;
+
+#include "memquery_test_data.h"
+
+memquery_library_bounds_test cur_bounds;
+int next_iter_position = -1;
+bool enable_memquery_unit_test;
+
+/* CHECK_EQ verifies that the provded x and y are equal; if not, it prints a
+ * message to STDERR and returns false.
+ */
+#define CHECK_EQ(x, y, fmt)                                                  \
+    if ((x) != (y)) {                                                        \
+        print_file(STDERR, #x " != " #y ": " fmt " != " fmt "\n", (x), (y)); \
+        return false;                                                        \
+    }
+
+/* Piggyback on CHECK_EQ for the failure message: since the strncmp
+ * showed the strings are not equal, we know that the x and y pointers are also
+ * not equal according to ==.
+ */
+#define ASSERT_STR_EQ(x, y, sz)         \
+    if (strncmp((x), (y), (sz)) != 0) { \
+        CHECK_EQ(x, y, "%s");           \
+    }
+
+bool
+run_single_memquery_test(memquery_library_bounds_test test)
```

Comment:
Done.

I'm guessing that's because there's really no reason to copy it here? That's sorta the same logic about the `cur_bounds` itself; do you think that should be a `memquery_library_bounds_test *` too?

Reply:
Right, these are constants, no need to memcpy all those fields.  Probably.

Reply:
Make cur_bounds a pointer too (since it let me easily put an ASSERT in these "test" functions to help catch future mistakes)